torch/csrc/jit/passes/peephole.cpp

```diff
@@ -15,368 +18,458 @@ static bool mustBeEqual(const c10::optional<T>& a, const c10::optional<T>& b) {
+              addmm_node->addInput(cOne);
+              addmm_node->addInput(cOne);
+              auto* addmm_value = addmm_node->output();
+
+              // Copy shape information from output node
+              addmm_value->copyMetadata(node->output());
+              GRAPH_UPDATE(
+                  "Fusing ",
+                  mm_node->input(0)->debugName(),
+                  ", ",
+                  mm_node->input(1)->debugName(),
+                  " and ",
+                  node->input(1 - mm_side)->debugName(),
+                  " into ",
+                  addmm_value->debugName());
+              node->output()->replaceAllUsesWith(addmm_value);
+              changed_ = true;
+              continue;
+            }
           }
         }
+        // TODO: this doesn't work with Scalar-Tensor ops! We should
+        // canonicalize those
       }
-      // TODO: this doesn't work with Scalar-Tensor ops! We should canonicalize
-      // those
-    } else if (
-        node->matches(
-            "aten::mul(Tensor self, Scalar other) -> Tensor",
-            /*const_inputs=*/attr::other) ||
-        node->matches(
-            "aten::div(Tensor self, Scalar other) -> Tensor",
-            /*const_inputs=*/attr::other)) {
-      // x * 1 == x / 1 == x
-      if (node->get<at::Scalar>(attr::other)->toDouble() == 1) {
-        GRAPH_UPDATE(
-            *node,
-            " (x * 1 == x / 1 == x) is replaced with ",
-            node->input(0)->debugName());
-        node->output()->replaceAllUsesWith(node->input(0));
-      }
-    } else if (
-        node->matches(
-            "aten::add(Tensor self, Scalar other, Scalar alpha) -> Tensor",
-            /*const_inputs=*/{attr::alpha, attr::other}) ||
-        node->matches(
-            "aten::sub(Tensor self, Scalar other, Scalar alpha) -> Tensor",
-            /*const_inputs=*/{attr::alpha, attr::other})) {
-      // x + 0 == x - 0 == x
-      if (node->get<at::Scalar>(attr::alpha)->toDouble() == 1 &&
-          node->get<at::Scalar>(attr::other)->toDouble() == 0) {
-        GRAPH_UPDATE(
-            *node,
-            " (x + 0 == x - 0 == x) is replaced with ",
-            node->input(0)->debugName());
-        node->output()->replaceAllUsesWith(node->input(0));
+
+      if (changed_) {
+        aliasDb_ = torch::make_unique<AliasDb>(graph_);
+        changed_ = false;
       }
-    } else if (
-        node->kind() == aten::Float || node->kind() == aten::Int ||
-        node->kind() == prim::ImplicitTensorToNum) {
-      Node* input_node = node->input()->node();
-      if (input_node->kind() == prim::NumToTensor) {
-        GRAPH_UPDATE(
-            *node,
-            " (x.NumToTensor().ImplicitTensorToNum() == x.NumToTensor()) is replaced with ",
-            node->input()->debugName());
-        node->output()->replaceAllUsesWith(input_node->input());
+
+      if(!aliasDb_->safeToChangeAliasingRelationship(node->inputs(), node->outputs())) {
```

Comment:
We're checking `safeToChangeAliasingRelationship` too generally here. There are only a specific set of transformations which we should be guarding with this call. For instance, we should not be guarding: 
`if an if node's output equals its condition replace output with condition` with this check.

Reply:
As far as I can tell, the only transformations that are introducing aliasing are:
`(x * 1 == x / 1 == x)` 
`(x + 0 == x - 0 == x)`

Reply:
Potentially `z + x.mm(y) == z.addmm(x, y) == x.mm(y) + z` however this is disabled atm

Reply:
> As far as I can tell, the only transformations that are introducing aliasing are:
(x * 1 == x / 1 == x)
(x + 0 == x - 0 == x)

there's more:
* `x-0`
* `x/1`
* `x.expand()`
* `x.t().t()`
* `unchecked_cast` for tensors, maybe? 

I think our default should be to run the slightly modified version of this check. 
This way when a new peephole is added, the check will block it and the author will have to scroll a little bit up in this file to actually learn a little bit about the dangers of aliasing. If they **still** think their transformation is safe to do they can whitelist it. Currently, the code is implicitly structured this way and I agree this isn't ideal. We should have all the transformations in one place and keep an exemption list elsewhere. If an operator is on the exemption list, we skip the check.
The modification to the check I'm thinking about is that we could also skip the check if the operator takes tensors/tensor lists but produces only primitives. This way if someone adds a peephole that extracts some tensor property or converts to a primitive, the modified check won't block such a case.

Reply:
`x-0 & x/1` were in the list I had above. 
The other ones don't apply: 
`x.expand(x.size()) is x == true`
`x.t()` is a view on the storage, so removing `x.t().t()` doesn't affect aliasing. 
`unchecked_cast` doesn't affect aliasing.

Reply:
I don't think we should be running logic that doesn't apply for all of our ops. If we're concerned that future authors will add unsafe aliasing, we can factor out the ops that introduce aliasing to a separate function and group them together that way.

Reply:
> x-0 & x/1 were in the list I had above.

Ah, yes, my bad, I didn't parse that properly 😄 

> If we're concerned that future authors will add unsafe aliasing, we can factor out the ops that introduce aliasing to a separate function and group them together that way.

I don't know if this refactoring gives us a strong enough guarantee that every unsafe op will indeed be added to this separate function. A future author might simply not notice this function and add their case to "safe" peephole ops. People use very different scanning strategies and tools when they read code. Some could easily go to the main function and search for the first `replaceAllUsesWith`, some might scroll top down and there's no good way to make sure this function won't be ignored no matter where we put it. It's unlikely that everyone will carefully read all the functions in this file since adding a new peephole seems so simple, which is exactly how we got this problem in the first place 😄 There are also code reviews, but again, relying on code reviews is far from being foolproof either.
On other hand, if they add their case and it doesn't work, they will have to understand why it doesn't work and dig deeper.

Reply:
We got into this problem in the first place because this code was written before mutation existed. It is generally true that future changes may be incorrect. Future optimizations may be incorrect for completely different reasons than unsafe alias introduction. Regardless of whether this check exist, reviewers and writers will have to access changes and their correctness conditions. This specific bug correctness condition applies to 2 out of the 14 or so optimizations we have in place, but we're now making it the general correctness condition for every optimization we make, potentially prevent valid transformations for the remaining 12. 

You have also written:
```
      // any transformations that are still valid despite
      // should be pulled above this statement
```
but haven't done so for any of the transformations that apply.